### ISSUE TITLE
Using Parallel Linq doesn't work with JSON serialisation

### ISSUE THREAD

**Issue by mmnosek** [Starcounter/Home#191]
I've got simple program:

```
namespace Test
{
    [Database]
    public class Entity
    {
        public string Something;
    }

    class Program
    {
        static void Main()
        {

            Db.Transact(() =>
            {
                for (var i = 0; i < 10; i++)
                {
                    new Entity { Something = i.ToString() };
                }
            });

            QueryResultRows<Entity> results = Db.SQL<Entity>("SELECT c FROM Test.Entity c");
            var results2 = from ent in results.AsParallel() where ent.Something == "aaa" select ent;
            var serialised = JsonConvert.SerializeObject(results2);
        }
    }
}
```

It causes following exception:
```
System.Exception: ScErrUnhandledException (SCERR1069): An unhandled exception was detected.
Version: 2.3.1.6870.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR1069.
   at StarcounterInternal.Hosting.Processor.RunTaskNoScheduling(Action userAction, String appName) in C:\TeamCity\BuildAgent\work\sc-6735\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 121
   at Starcounter.DbSession.GetAndExecuteQueuedTasks(Action`2 taskExecutionMethod) in C:\TeamCity\BuildAgent\work\sc-6735\Level1\src\Starcounter\DbSession.cs:line 194
   at StarcounterInternal.Hosting.Processor.RunTask(IntPtr hTask) in C:\TeamCity\BuildAgent\work\sc-6735\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 168
   at StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) in C:\TeamCity\BuildAgent\work\sc-6735\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 81
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR1069
---> System.AggregateException: One or more errors occurred.
   at System.Linq.Parallel.QueryTaskGroupState.QueryEnd(Boolean userInitiatedDispose)
   at System.Linq.Parallel.AsynchronousChannelMergeEnumerator`1.MoveNextSlowPath()
   at System.Linq.Parallel.QueryOpeningEnumerator`1.MoveNext()
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeList(JsonWriter writer, IEnumerable values, JsonArrayContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.Serialize(JsonWriter jsonWriter, Object value, Type objectType)
   at Newtonsoft.Json.JsonSerializer.SerializeInternal(JsonWriter jsonWriter, Object value, Type objectType)
   at Newtonsoft.Json.JsonConvert.SerializeObjectInternal(Object value, Type type, JsonSerializer jsonSerializer)
   at Newtonsoft.Json.JsonConvert.SerializeObject(Object value)
   at Test.Program.<>c__DisplayClass0_0.<Main>b__2() in C:\Projects\Test\Test\Program.cs:line 33
   at Starcounter.DbSession.<>c__DisplayClass5_0.<RunAsync>b__0() in C:\TeamCity\BuildAgent\work\sc-6735\Level1\src\Starcounter\DbSession.cs:line 216
   at Starcounter.Task.Run() in C:\TeamCity\BuildAgent\work\sc-6735\Level1\src\Starcounter\DbSession.cs:line 100
HResult=-2146233088
---> System.Exception: ScErrThreadNotAttached (SCERR1004): The operation failed because the current thread wasn't attached to the scheduler.
Version: 2.3.1.6870.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR1004.
   at Starcounter.ThreadData.GetContextHandleExcept() in C:\TeamCity\BuildAgent\work\sc-6735\Level1\src\Starcounter\Internal\ThreadData.cs:line 70
   at Starcounter.IteratorHelper.GetEnumeratorCached_NoCodeGenFilter(UInt32 rangeFlags, Byte[] firstKey, Byte[] lastKey, Enumerator cachedEnum) in C:\TeamCity\BuildAgent\work\sc-6735\Level1\src\Starcounter\Mock\IteratorHelper.cs:line 168
   at Starcounter.Query.Execution.FullTableScan.CreateEnumeratorWithoutNativeFilter() in C:\TeamCity\BuildAgent\work\sc-6735\Level1\src\Starcounter\Query\Execution\Enumerators\FullTableScan.cs:line 396
   at Starcounter.Query.Execution.FullTableScan.MoveNext() in C:\TeamCity\BuildAgent\work\sc-6735\Level1\src\Starcounter\Query\Execution\Enumerators\FullTableScan.cs:line 411
   at System.Linq.Parallel.PartitionedDataSource`1.ContiguousChunkLazyEnumerator.MoveNext(T& currentElement, Int32& currentKey)
   at System.Linq.Parallel.WhereQueryOperator`1.WhereQueryOperatorEnumerator`1.MoveNext(TInputOutput& currentElement, TKey& currentKey)
   at System.Linq.Parallel.PipelineSpoolingTask`2.SpoolingWork()
   at System.Linq.Parallel.SpoolingTaskBase.Work()
   at System.Linq.Parallel.QueryTask.BaseWork(Object unused)
   at System.Threading.Tasks.Task.Execute()
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR1004
```

Changing
```
var results2 = from ent in results.AsParallel() where ent.Something == "aaa" select ent;
```
to
```
var results2 = from ent in results where ent.Something == "aaa" select ent;
```

Works perfectly fine. Why using Parallel Linq causes this exception?:

**Comment by bigwad** [comment]
Because linq iterates sequence (thus effectively querying the database) in thread pool, which is not compatible with starcounter scheduling.

**Comment by miyconst** [comment]
Closing the issue as answered question.

**Comment by atifwaqar** [comment]
Hi @mmnosek & @miyconst , Is following exception same as above?

```
System.Exception: ScErrUnhandledException (SCERR1069): An unhandled exception was detected.
Version: 2.3.1.7166.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR1069.
   at StarcounterInternal.Hosting.Processor.<>c__DisplayClass3_0.<RunTaskNoScheduling>b__0() in C:\TeamCity\BuildAgent\work\sc-8192\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 138
   at Starcounter.Internal.StarcounterEnvironment.RunWithinApplicationInternal(String desiredAppName, Action action) in C:\TeamCity\BuildAgent\work\sc-8192\Level1\src\Starcounter.Internal\Mock\StarcounterEnvironment.cs:line 142
   at StarcounterInternal.Hosting.Processor.RunTaskNoScheduling(Action userAction, String appName) in C:\TeamCity\BuildAgent\work\sc-8192\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 121
   at Starcounter.DbSession.GetAndExecuteQueuedTasks(Action`2 taskExecutionMethod) in C:\TeamCity\BuildAgent\work\sc-8192\Level1\src\Starcounter\DbSession.cs:line 180
   at StarcounterInternal.Hosting.Processor.RunTask(IntPtr hTask) in C:\TeamCity\BuildAgent\work\sc-8192\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 168
   at StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) in C:\TeamCity\BuildAgent\work\sc-8192\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 81
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR1069
---> System.Exception: ScErrAcquireSessionTimeout (SCERR5008): Exclusive access to the session could not be obtained. Id: 74C5A3CC7EBD0B4000000040
Version: 2.3.1.7166.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR5008.
   at Starcounter.Session.StartUsing() in C:\TeamCity\BuildAgent\work\sc-8192\Level1\src\Starcounter.XSON\Sessions\Session.cs:line 614
   at Starcounter.Session.Use(SessionTask action, String sessionId) in C:\TeamCity\BuildAgent\work\sc-8192\Level1\src\Starcounter.XSON\Sessions\Session.cs:line 797
   at Starcounter.DbSession.<>c__DisplayClass5_0.<RunAsync>b__0() in C:\TeamCity\BuildAgent\work\sc-8192\Level1\src\Starcounter\DbSession.cs:line 216
   at Starcounter.Task.Run() in C:\TeamCity\BuildAgent\work\sc-8192\Level1\src\Starcounter\DbSession.cs:line 100
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR5008
```

**Comment by miyconst** [comment]
No, those are different issues.

**Comment by atifwaqar** [comment]
Is it fixed in some later SC version? 
My current version is 2.3.1.7166

**Comment by miyconst** [comment]
No. This issue will not be fixed in `2.3.x`.

**Comment by atifwaqar** [comment]
Alright.
Just out of curiosity, will this be fixed in 2.4? and also, is there any GH Issue created where I can follow up on this issue?

**Comment by miyconst** [comment]
It is not planned to be fixed in `2.4` either, since it requires to implement new/different scheduling model in order to be able to access database from standard .Net thread. 

We already have some plans, but when and how is not yet decided, since there are other more important tasks.

**Comment by gijswijs** [comment]
@bigwad Just wanted to share my gratitude. Your remark on when and where Linq iterates put me on the path of the solution of a related problem in my own project. It had me stumped for more than a day. Thanks a million.
